packages/runed/src/lib/utilities/useNetworkStatus/useNetworkStatus.svelte.ts

```diff
@@ -0,0 +1,145 @@
+	 * underlying connection technology
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/downlinkMax
+	 */
+	downlinkMax?: NetworkInformation["downlinkMax"];
+	/**
+	 * @desc Effective type of the connection meaning one of 'slow-2g', '2g', '3g', or '4g'.
+	 * This value is determined using a combination of recently observed round-trip time
+	 * and downlink values.
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/effectiveType
+	 */
+	effectiveType?: NetworkInformation["effectiveType"];
+	/**
+	 * @desc Estimated effective round-trip time of the current connection, rounded
+	 * to the nearest multiple of 25 milliseconds
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/rtt
+	 */
+	rtt?: NetworkInformation["rtt"];
+	/**
+	 * @desc {true} if the user has set a reduced data usage option on the user agent.
+	 * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/saveData
+	 */
+	saveData?: NetworkInformation["saveData"];
+	/**
+	 * @desc The type of connection a device is using to communicate with the network.
+	 * It will be one of the following values:
+	 *  - bluetooth
+	 *  - cellular
+	 *  - ethernet
+	 *  - none
+	 *  - wifi
+	 *  - wimax
+	 *  - other
+	 *  - unknown
+	 *  @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation/type
+	 */
+	type?: NetworkInformation["type"];
+}
+
+/**
+ * Tracks the state of browser's network connection.
+ * @returns null if the function is not run in the browser
+ */
+export function useNetworkStatus() {
+	if (!browser) {
+		return {
+			get current() {
+				return null;
+			},
+			get previous() {
+				return null;
+			},
+		};
+	}
+	const navigator = window.navigator;
+	const connection =
+		"connection" in navigator ? (navigator as NavigatorWithConnection).connection : undefined;
+	const getConnectionStatus = (): NetworkStatus => {
+		return {
+			online: navigator.onLine,
+			updatedAt: new Date(),
+			downlink: connection?.downlink,
+			downlinkMax: connection?.downlinkMax,
+			effectiveType: connection?.effectiveType,
+			rtt: connection?.rtt,
+			saveData: connection?.saveData,
+			type: connection?.type,
+		};
+	};
+
+	let current = $state(getConnectionStatus());
+	const previous = new Previous(() => current);
+	const handleStatusChange = () => {
+		current = getConnectionStatus();
+	};
+
+	$effect(() => {
+		// The connection event handler also manages online and offline states.
+		if (connection) {
+			addEventListener(connection, "change", handleStatusChange, { passive: true });
+		} else {
+			addEventListener(window, "online", handleStatusChange, { passive: true });
+			addEventListener(window, "offline", handleStatusChange, { passive: true });
+		}
+
+		return () => {
+			window.removeEventListener("online", handleStatusChange);
+			window.removeEventListener("online", handleStatusChange);
+			connection?.removeEventListener("change", handleStatusChange);
+		};
+	});
```

Comment:
`addEventListener` returns a cleanup function, so you can just rework to something like this.

```suggestion
	$effect(() => {
		const callbacks: VoidFunction[] = [];

		// The connection event handler also manages online and offline states.
		if (connection) {
			callbacks.push(addEventListener(connection, "change", handleStatusChange, { passive: true }));
		} else {
			callbacks.push(addEventListener(window, "online", handleStatusChange, { passive: true }));
			callbacks.push(addEventListener(window, "offline", handleStatusChange, { passive: true }));
		}

		return () => {
			callbacks.forEach((c) => c());
		};
	});
```

Reply:
Thanks, this is useful. I missed that `addEventListener` returns a cleanup function. 
Refactored it.